compiler/src/iree/compiler/Preprocessing/Common/test/attr_based_pipeline.mlir

```diff
@@ -1,21 +1,54 @@
     preprocessing_pipeline = #util.preprocessing_pipeline<"iree-preprocessing-make-single-dispatch">} {
-  %transposed = linalg.transpose ins(%lhs : tensor<16x26x18x288xbf16>) outs(%outs : tensor<16x288x26x18xbf16>) permutation = [0, 3, 1, 2]
+  %transposed = linalg.transpose ins(%lhs : tensor<1x26x18x288xbf16>) outs(%outs : tensor<1x288x26x18xbf16>) permutation = [0, 3, 1, 2]
   %conv = linalg.conv_2d_nchw_fchw {dilations = dense<1> : vector<2xi64>, strides = dense<1> : vector<2xi64>}
-    ins(%transposed, %rhs : tensor<16x288x26x18xbf16>, tensor<288x288x3x3xbf16>)
-    outs(%outs2 : tensor<16x288x24x16xf32>) -> tensor<16x288x24x16xf32>
-  return %conv : tensor<16x288x24x16xf32>
+    ins(%transposed, %rhs : tensor<1x288x26x18xbf16>, tensor<288x288x3x3xbf16>)
+    outs(%outs2 : tensor<1x288x24x16xf32>) -> tensor<1x288x24x16xf32>
+  return %conv : tensor<1x288x24x16xf32>
 }
-// CHECK-LABEL: test
-//  CHECK-SAME: %[[ARG0:[A-Za-z0-9]+]]: tensor<16x26x18x288xbf16>
+// CHECK-LABEL: @single_dispatch_dropunitdims
+//  CHECK-SAME: %[[ARG0:[A-Za-z0-9]+]]: tensor<1x26x18x288xbf16>
 //       CHECK:   %[[DISPATCH:.+]] = flow.dispatch.region
-//       CHECK:     %[[CONV:.+]] = linalg.generic {{.*}} ins(%[[ARG0]]
+//       CHECK:     %[[COLLAPSE:.+]] = tensor.collapse_shape %[[ARG0]]
+//       CHECK:     %[[EXPAND:.+]] = tensor.expand_shape %[[COLLAPSE]]
+//       CHECK:     %[[CONV:.+]] = linalg.generic {{.*}} ins(%[[EXPAND]]
 //       CHECK:     flow.return %[[CONV]]
 //       CHECK:   return %[[DISPATCH]]
 
+// -----
+func.func @single_dispatch_fusion(%lhs : tensor<18x288xf32>, %rhs :  tensor<18x288xbf16>, %outs : tensor<18x288xbf16>)
+    -> tensor<18x288xbf16> attributes {
+    preprocessing_pipeline = #util.preprocessing_pipeline<"iree-preprocessing-make-single-dispatch">} {
+  %first = linalg.generic {
+    indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>,
+                     affine_map<(d0, d1) -> (d0, d1)>],
+    iterator_types = ["parallel", "parallel"]}
+    ins(%lhs : tensor<18x288xf32>) outs(%outs : tensor<18x288xbf16>) {
+  ^bb0(%in: f32, %out: bf16):
+    %1 = arith.truncf %in : f32 to bf16
+    linalg.yield %1 : bf16
+  } -> tensor<18x288xbf16>
+  %final = linalg.generic {
+    indexing_maps = [affine_map<(d0, d1) -> (d0, d1)>,
+                     affine_map<(d0, d1) -> (d0, d1)>,
+                     affine_map<(d0, d1) -> (d0, d1)>],
+    iterator_types = ["parallel", "parallel"]}
+    ins(%first, %rhs : tensor<18x288xbf16>, tensor<18x288xbf16>) outs(%outs : tensor<18x288xbf16>) {
+  ^bb0(%in: bf16, %in_3: bf16, %out: bf16):
+    %2 = arith.addf %in, %in_3 : bf16
+    linalg.yield %2 : bf16
+  } -> tensor<18x288xbf16>
+  return %final : tensor<18x288xbf16>
+}
+
+// CHECK-LABEL: @single_dispatch_fusion
+//       CHECK:   %[[DISPATCH:.+]] = flow.dispatch.region
+//       CHECK:   %[[TRUNC:.+]] =  arith.truncf
+//       CHECK:   %[[ADD:.+]] = arith.addf %[[TRUNC]]
```

Comment:
I think it will be more clear if you check these ops within the linalg.generic.

Reply:
Yes transpose propagation can add back transposes and we need to generalize them again. The default path also does this in dispatch creation.